### PR TITLE
add 3D textures

### PR DESCRIPTION
**Pull request by karroffel**
This PR adds new texture types, `TEXTURE_TYPE_2D_ARRAY` and `TEXTURE_TYPE_3D`.

The `VisualServer` API is changed slightly to compensate for the new dimensions needed. (@reduz said on IRC that this is acceptable since the servers are rarely used in games and are mostly for internal use)

The new types `sampler3D`, `isampler3D`, `usampler3D` as well as `sampler2DArray`, `isampler2DArray` and `usampler2DArray` have been added to the shader language, accompanied by the appropriate accessor functions.

A "highlevel" resource type, `Texture3D` has been added that can be used in-editor as a shader parameter by *splitting an image into many layers*. Via code the `Texture3D` can be filled with arbitrary data from differing formats.

Later on I will add a `Texture2DArray` highlevel resource type for completion.

### CODE REVIEW

**Comment by Zylann** [comment]
Great stuff!
I am curious how users (or plugins) will be able to create such textures without needlessly importing 2D textures

**Comment by karroffel** [comment]
@Zylann the `Texture3D` type accepts `Ref<Image>` for the data, I am using that for the property as well.

That's how the editor-interface looks like:


Juan pushed a commit a day ago to import image files as `Image` and not as `Texture`, so they don't reside in VRAM. This PR makes use of that.

So any texture you set as the "backend" doesn't have to be uploaded to GPU, so that should help a lot. (If I understood your concerns correctly)

**Comment by Two-Tone** [comment]
This could be useful in the medical industry. Stuff like fMRI scans are stored as just a bunch of textures (at least mine were).

**Comment by karroffel** [comment]
@Two-Tone always a step ahead of you :)

https://gfycat.com/FluidNeatAfghanhound

**Comment by Two-Tone** [comment]
Sweet. Although I am curious as to what's up with the wobbling in the
texture as the camera rotates.

**Comment by karroffel** [comment]
@Two-Tone that's a problem with the shader I had, now I have a better shader that does perspective correction, that was just a Proof-of-concept, don't worry, the 3D texture works fine and has no wobbling :D

**Comment by Two-Tone** [comment]
That's great to hear! If I can find my copies of my scans, maybe I'll try
to throw them into Godot and look at them from a tool that doesn't make me
want to tear my eyes out.

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 1]
```diff
@@ -1247,6 +1226,59 @@ bool RasterizerSceneGLES3::_setup_material(RasterizerStorageGLES3::Material *p_m
...
+							tex = storage->resources.white_tex_3d;
+						} break;
+					}
+
+				} break;
+
+				case ShaderLanguage::TYPE_ISAMPLER2DARRAY:
+				case ShaderLanguage::TYPE_USAMPLER2DARRAY:
+				case ShaderLanguage::TYPE_SAMPLER2DARRAY: {
+					// TODO
```
[the remark this answers is not in the document]
Comment (karroffel) [reply drivers/gles3/rasterizer_scene_gles3.cpp]: Yeah as I said, I want to add a `Texture2DArray` resource later on, just wanted to get this out and it already works if you use the low-level RIDs, but I'll definitely complete thsi! :+1:

File: scene/resources/texture.cpp [hunk 2]
```diff
@@ -1876,3 +1876,209 @@ AnimatedTexture::AnimatedTexture() {
 AnimatedTexture::~AnimatedTexture() {
 	VS::get_singleton()->free(proxy);
 }
+
+bool Texture3D::get_split_single_image_enabled() const {
```
[the remark this answers is not in the document]
Comment (karroffel) [reply scene/resources/texture.cpp]: Okay, I don't know how to deal with importers, so that would be great!

**Pull request merged**